### ISSUE TITLE
Language of Program vs website vs listed as supported

### ISSUE THREAD

**Issue by MBB232** [OpenShot/openshot-qt#3406]
There seems to be inconsistency between which languages are supported. 
On the web interface of openshot.org 7 languages are listed
On the supported languages list Support>luanguage, only 3 languages are listed
https://code.launchpad.net/~openshot.code/openshot/translations-2.0
(Webside may link to old repository here)

When actually using openshot, a **very** long list of languages can be selected

**Describe the bug:**

**Steps to reproduce the behavior:**
1. Go to 'Features-language-Dutch (or French) 
2. Click on 'support>User guide'
3. User guide remains in English

**Expected behavior: (of website)**
- Have a button with flag icons in the top corner to select language
- See only languages listed that are fully translated.
- Expect the manual to be in selected language (Dutch/French/?)

**System Details:**
- Chrome + Edge

**Comment by ferdnyc** [comment]
Hey @MBB232, thanks for bringing this up. I'm a bit confused about a few things you mentioned... 

> There seems to be inconsistency between which languages are supported.

It's likely there will be _disparities_, at least, yeah.

> On the web interface of openshot.org 7 languages are listed

![Screenshot from 2020-04-24 14-46-31](https://user-images.githubusercontent.com/538020/80246901-2cbd0900-863b-11ea-8915-fa77904dcfd0.png)

I count 18, but for sure it's a shorter list than in OpenShot itself. That makes sense, though, as that list is a list of languages the **website** is translated into. OpenShot-the-application supports many more languages than that, thanks to the excellent community of volunteer translators who've contributed via the launchpad project.

> On the supported languages list Support>luanguage, only 3 languages are listed
> https://code.launchpad.net/~openshot.code/openshot/translations-2.0
> (Webside may link to old repository here)

That part confuses me, because I see a _very_ long list. There are only three that are **fully** translated, but there are many dozens more with at least partial translations:

![image](https://user-images.githubusercontent.com/538020/80246572-a1437800-863a-11ea-8214-2f494eac947a.png)

It continues quite a ways beyond the bottom of my browser window. Possibly you need to adjust your launchpad settings, and/or log in. Could be that it only shows fully translated languages by default?

...Or am I misunderstanding you?

**Comment by ferdnyc** [comment]
Oh, I've got you. You're talking about the _user guide_ languages. Yeah, I don't know whether the documentation has been translated at all, ISTR an old issue mentioning that it was only available in English currently. I'll see if I can look that up.

**Comment by ferdnyc** [comment]
Nope, can't find it anywhere. Oh well.

The long and short of it seems to be that I don't believe we currently have a method of extracting text from the reStructuredText documentation sources (including trickier things like table contents and image captions, to say nothing of the actual images themselves), or of getting that text into a form that could be uploaded to launchpad. And I know for a _fact_ we don't have any mechanism in place to generate the HTML docs with the theoretical translations applied, even if they did exist.

Unfortunately, as things currently stand the User Guide seems stuck as a single-language affair, though ideas for changing that would definitely be welcomed.

**Comment by MBB232** [comment]
> > On the web interface of openshot.org 7 languages are listed
> 
> ![Screenshot from 2020-04-24 14-46-31](https://user-images.githubusercontent.com/538020/80246901-2cbd0900-863b-11ea-8915-fa77904dcfd0.png)
> 
> I count 18, but for sure it's a shorter list than in OpenShot itself. 

My bad; browsers cut off that list when it reaches the end of the screen (or perhaps it does not, it keeps running down in your screenshot?)
So on a 15" laptop screen with zoom to 110%, it only shows 7, on a larger screen 9, etc. 
Because there was space left above the list, I did not realize that it was being cut off. 

I still would recommend the enhancement to have "Language" as a main option in the top bar, rather then hidden under Features.

<img width="840" alt="Website_ZoomScreenCutOff" src="https://user-images.githubusercontent.com/62191951/80272281-393b7300-86c8-11ea-8aac-e5d7d61de34c.PNG">


> > On the supported languages list Support>luanguage, only 3 languages are listed
> > https://code.launchpad.net/~openshot.code/openshot/translations-2.0
> > (Webside may link to old repository here)
> 
> That part confuses me, because I see a _very_ long list. There are only three that are **fully** translated, but there are many dozens more with at least partial translations:
> > 
> It continues quite a ways beyond the bottom of my browser window. Possibly you need to adjust your launchpad settings, and/or log in. Could be that it only shows fully translated languages by default?
> 

Again, my bad; I had not seen the "List all languages" link. (I can not sign in to change preferences because I do not have an account)

By default it seems to show only languages based on installed region or IP something; it shows Dutch, Frisian (Dutch) and German on one machine (two different browsers), and only Dutch and Frisian on the other. 
It is actually kind of creepy as I did not know I was leaking such region info. (OS and browsers are both in English; I do have German spellcheck installed in the browser)

![WebsiteLink_3languages](https://user-images.githubusercontent.com/62191951/80272288-48babc00-86c8-11ea-8aca-3870c45ec099.PNG)

Weirder still is that the link that I posted above myself by cut-and-past is somehow changed, and now lists no languages at all. Perhaps because I used one of the connect buttons in this editor

**Comment by MBB232** [comment]
> The long and short of it seems to be that I don't believe we currently have a method of extracting text from the reStructuredText documentation sources (including trickier things like table contents and image captions, to say nothing of the actual images themselves), or of getting that text into a form that could be uploaded to launchpad. And I know for a _fact_ we don't have any mechanism in place to generate the HTML docs with the theoretical translations applied, even if they did exist.
> 
> Unfortunately, as things currently stand the User Guide seems stuck as a single-language affair, though ideas for changing that would definitely be welcomed.

Ah right, that would explain it. Seeing languages listed as 100% translated led me incorrectly to believe that the manual was done too. 

A quick search said that Sphinx itself should be able to extract and re-insert Po(t) files from .rst?
http://www.sphinx-doc.org/en/1.5.1/intl.html

OmegaT won't recognize .rst unless I rename them to .txt, although there seem to be attempts at plugins on github. 

I can't get them to load into KDE Lokalize either even though it uses them in its own documentation. But that is probably my complete misunderstanding of how the program works as it has been fighting me every step. 

I see how tables would be an issue considering the importance of spaces in Rest, but a short explanation could make translators fix that - still easier then rewriting without tools. 

As for making images for every language, that would indeed be a lot of work. But you said before that all actions on images should be in the description anyway, so using the English image for a vague indication would probably work well enough.

**Comment by ferdnyc** [comment]
Yeah, I took a quick look at Sphinx's i18n features tonight. It's actually pretty workable. Extracting gettext translation masters is a cinch, and generating multiple versions of the documentation is just a matter of dropping the files in the right directory and scripting Sphinx to build N different translated copies of the docs.

German-localized versions of images can be stored as <code><var>filename</var>.de.png</code>, or as <code>images/de/<var>filename</var>.png</code> -- or basically whatever else we come up with, there's a variable in the config file where you can define a template it'll use to generate filenames for each language.

The translation master is a bit raw, you end up with individual table cell contents like "2" and "Save" getting pulled out as translatable strings with no real context other than a line number in an `.rst` source file. If there's a way to attach translator notes at the source to provide some context, and have them get included in the translation master, I haven't seen any example or mention of it yet. But there probably is, that's a pretty fundamental gettext feature.

**Comment by ferdnyc** [comment]
I'm not sure how easy it would be to get translated versions of the manual hooked up on the OpenShot.org website, or what would be required to make that work — that's all 100% @jonoomph's domain.

But first step would be to get people the masters over @ launchpad so they can begin submitting translations, once we had those we could at least start producing translated local documentation for the installers and Linux packages.

So, actually, _first_ first thing would be to make whatever changes need to be made to the docs, because it would be shitty to post a bunch of files for translation just to throw away half of the work people do when we edit the source material.

Which actually reminds me of something else I wanted to suggest: They're not written this way currently, but there's a convention that started in the ASCIIdoc world — it tends to be a pretty good fit for any sort of written prose, when it's in a markup language like adoc/reST/MarkDown and managed in version control: Sentence-per-line.

It's just what it sounds like: You start each new sentence on its own line in the source, and let the markup language flow them all together into a paragraph when it generates the formatted docs. (You can also break at other logical points, like after a comma in a longer sentence, or before starting an inline markup command. It's a guideline, not a rule.)

Writing that way, there's no worrying about line length or when to wrap, but at the same time it encourages shorter, simpler sentences which is a good thing when writing docs. The diffs when changes are submitted also tend to be more readable and focused.

Based on the experiences I've had so far with writing collaborative docs that way, I'm a convert. I definitely recommend trying it out, and I'll most likely be reformatting parts of the user guide source that way when I happen to touch them.

**Comment by MBB232** [comment]
Docs/de/images may be best, then the translated files have their own directory to live in. Otherwise you may need to translate filenames and break links. 

So Disable wordwrap and run on till the end of the sentence. 
I actually had a question about that. In normal digital writing it i. s habit to end a sentence on a space after the dot. But current documentation does not do that. Are spaces seen as hard return in rest?

I'll do a demo of it in my first complete update so we can see the difference. 

I'll also add it to my collection of docs_about.rst
(NB: still waiting the in docs topic for your actual permission to re-use your comments here in , as per your own statement that it should not be assumed. )

PS: can you add the tag 'docs' to this topic?

**Comment by ferdnyc** [comment]
> PS: can you add the tag 'docs' to this topic?

Sure, I also added it to the docs-update Project. I pulled "bug", though, because to the extent these are openshot.org site bugs, those aren't tracked here. :grin: (They aren't tracked _anywhere_, formally, but point is there's no OpenShot application bug involved.)

**Comment by ferdnyc** [comment]
> My bad; browsers cut off that list when it reaches the end of the screen (or perhaps it does not, it keeps running down in your screenshot?)
> So on a 15" laptop screen with zoom to 110%, it only shows 7, on a larger screen 9, etc.
> Because there was space left above the list, I did not realize that it was being cut off.

Aha! Whereas I have a 1920×1200 screen, so I never realized it couldn't be scrolled up if it was longer than the window. But now I see that that's the case. Must be the site template. `¯\_(ツ)_/¯`

**Comment by ferdnyc** [comment]
@MBB232 

> Docs/de/images may be best, then the translated files have their own directory to live in. Otherwise you may need to translate filenames and break links.

Oh, that's all handled by Sphinx. When it generates the localized documentation, it automatically replaces references (including image filenames) with translated versions if they exist, otherwise it leaves the original in place. We can use any system we like, we just have to tell it what that system is.

> So Disable wordwrap and run on till the end of the sentence.

Well, it's more the other way around: **Don't** continue past the end of a sentence or other logical block, start each one on a new line.

Prose content is normally written in paragraph-sized units. They may be automatically or manually wrapped, but organizationally each sentence immediately follows the end of the previous one.

Most markup systems, though — with the notable (and annoying) exception of GitHub comments — will treat immediately consecutive lines as part of the same paragraph, and automatically flow them together.

Which means that, though the source code is formatted differently, these two blocks of text are exactly equivalent in reStructuredText:
```rst
OpenShot is a video editor for Linux. It is written in Python. You can
download it from https://openshot.org/download.
```
```rst
OpenShot is a video editor for Linux.
It is written in Python.
You can download it from https://openshot.org/download.
```

The advantage of sentence-per-line is in what happens when the text is subsequently edited. Say I want to change only the first sentence of that paragraph. If I edit the first block and submit a pull request, the content diff would probably look like this:
```diff
diff --git a/paragraph.rst b/paragraph.rst
index 7f0f7ef..58c39e5 100644
--- a/paragraph.rst
+++ b/paragraph.rst
@@ -1,2 +1,3 @@
-OpenShot is a video editor for Linux. It is written in Python. You can
-download it from https://openshot.org/download.
+OpenShot Video Editor is a cross-platform application that runs on Linux,
+Windows, and macOS. It is written in Python. You can download it from
+https://openshot.org/download.
```

Whereas if I make the same edit to the second block — even with wrapping turned on, because `vim` automatically wraps .rst files at 80 characters and I didn't feel like turning it off for this example — the resulting diff looks like this:
```diff
diff --git a/sentence.rst b/sentence.rst
index 672c901..24f1b7c 100644
--- a/sentence.rst
+++ b/sentence.rst
@@ -1,3 +1,4 @@
-OpenShot is a video editor for Linux.
+OpenShot Video Editor is a cross-platform application that runs on Linux,
+Windows, and macOS.
 It is written in Python.
 You can download it from https://openshot.org/download.
```
It's a _much_ shorter diff, and it's immediately clear that the first sentence is the only one I edited, something that's completely hidden in the wrapped-paragraph diff.

> I actually had a question about that. In normal digital writing it i. s habit to end a sentence on a space after the dot. But current documentation does not do that. Are spaces seen as hard return in rest?

I'm not sure I follow, but I may have just inadvertently answered your question. For the most part ReST outputs content exactly as written, so handling of whitespace is governed by how HTML processes it. In fact, when I said earlier that ReST reflows consecutive lines into paragraphs, that's not actually true. It outputs the lines as written, on multiple lines in the HTML source. But they're all part of the same `<p>...</p>` (paragraph) block, and _HTML_ will reflow a paragraph together whether it's written on one line or 20 — all that matters is that all of the lines are contained in the same paragraph block.

And in HTML, all whitespace is equivalent — space, tab, newline, doesn't matter — it's all "a space", logically. Multiple consecutive whitespace characters are compressed into a single logical content space. So all three of these are exactly equivalent:
```html
<p>OpenShot Video Editor</p>

<p>OpenShot         Video                               Editor</p>

<p>
OpenShot
Video
Editor
</p>
```

They'll all format as:
<p>OpenShot Video Editor</p>

The only difference is that in ReST, you separate content blocks with two consecutive newlines. When processing the source, ReST wraps each block in its own set of `<p>...</p>` tags. So, unlike in HTML, in ReST this  _isn't_ an equivalent form:
```rst
OpenShot Video

Editor
```
because it'll be turned into this HTML:
```html
<p>OpenShot Video</p>
<p>Editor</p>
```
But if we were writing raw HTML, this would be **exactly** equivalent to the earlier set of examples:
```html
<p>OpenShot

                      Video

Editor</p>
```

A lot of the existing documentation content uses two spaces after a period, because @jonoomph was likely abused by his educational system, which ingrained him with rules that should have been left in the age of manual typewriters. :wink: There's no _harm_ in typing two spaces after a period, but also no _point_ — in HTML it'll still render as a single space. Nevertheless, it's habit for a lot of people.

A far better habit, as I've demonstrated, is to begin each sentence on a new line.

**Comment by ferdnyc** [comment]
(If you're wondering why I mentioned GitHub as an exception to the reflow rule:
They made the unfortunate decision to replace each newline in comment text with a `<br/>` tag.
So, you type a newline, you get a new line.
That sounds like a good idea at first, but in practice it creates a lot of problems.
Including the fact that this comment looks like this, because I wrote it sentence-per-line.)

**Comment by ferdnyc** [comment]
> PS2: would table issues be solved by always specifying table width?

See #1262 / #1272 for the hows and whys of column widths — basically, column widths are manually specified because I inserted CSS to enable text wrapping in table cells. (Normally in the ReadTheDocs theme they don't wrap at all, under any circumstances, which is just terrible. And if you're on a mobile device, unreadable.)

Once wrapping is enabled, you often have to manually adjust the balance of column widths to get _sensible_ wrapping. Otherwise the widths are scaled proportionally so wider columns completely overpower narrower ones, when the sensible thing would be to wrap only the wide column (if possible, given the available space).

Beyond that, there are no table issues that I'm aware of. If you mean the issue with _translating_ table cell contents, that's just down to how translations are managed by gettext.

#### A not-at-all brief primer on software localization
As gettext was primarily designed for translating software, the typical unit of translation is a string literal. That's done both to make it possible to easily replace elements with their translated versions, since you can usually swap out one string for another without changing how the program compiles, _and_ it's done because the same strings will often being repeated multiple times throughout the code, which often means they can all have the same translation.

That's not _always_ the case, though. In fact, when a string is a single word or short phrase, that string can have very different meanings depending on the context where it's used.

(As an extreme — and extremely contrived — example, imagine a document editor uses the label "Insert". That same text could appear in different places to mean very different things: an edit mode, various actions (inserting a piece of text, inserting a row or column in a table editor), and so on. The same term may apply to all of those uses in English, but when the UI is translated to another language, there could be a completely different word for each of them. It all depends on the context.)

Because of that problem, most translation systems support some notion of _translation context_. Qt's QTranslate, for example, makes each UI file (== window-or-dialog, typically) a separate context: "Advanced" in our main window's View menu isn't assumed to mean the same thing as "Advanced" in our Export dialog. And within the context, because it's a GUI localization system, the QLinguist software presents each context visually: The UI template is displayed, and each button/field/etc. to be labeled is shown in the context of the overall interface.

Gettext is used for translating all types of programs, including those without a GUI, or any UI at all. (It may even predate most modern GUI systems.) So, it contextualizes by source file reference, and supports translator comments for additional context. In a typical application, localizable strings are run through a translation function, named `_` by convention. Anywhere `_("Advanced")` appears in the source code, gettext will extract the contents as a translatable string in the `.pot` (translation master) file, marked with the line number or numbers where it was encountered.

When translators localize an application, they create a `.po` file, which is just the `.pot` file with some or all of the source strings translated into a given target language. If all of the uses of a string are the same, it only has to be translated once. When the program is run in the given target language, its corresponding `.po` data is loaded. Each time there's a `_()` call, if a translation is available it will be used instead of the original string.

Really it's quite  a bit more complex than that. (in part because languages have vastly different rules for pluralization, so any text with a countable subject has to take the quantity into account for proper translation. Things like date/time values also have to be treated specially. Even things like numeric values are formatted differently in different locales — `1,437.65` vs. `1.437,65` for example.) But that's the gist of it.

#### Translator comments
Because even all of those features don't ensure that every string will be easily translatable, gettext also supports manual contextualization using translator comments. When it processes this Python source code:
```python
# TRANSLATORS: A Help menu option to load the user guide website
menuHelp.addMenu(_("Open Manual"))
```
Because the comment immediately precedes the translatable text, gettext will attach that note as a comment on the `"Open Manual"` string in the `.pot` file. Otherwise, translators would see only the string itself, and without additional context the phrase `"Open Manual"` is completely ambiguous. (They'd also have the source code line number, but translators aren't expected to be fluent in a program's source language. Even if they are, `menuHelp.addMenu(_("Open Manual"))` is only marginally more informative. You'd have to read a lot more of the code to find out what that menu option _does_.)

#### Translation in Sphinx
Unlike most source code languages, ReStructuredText doesn't allow arbitrary comments to be inserted anywhere you like. You can only use block comments, and those have to be separate from content blocks — by their very definition, comments _can't_ be attached to content elements, especially any that are smaller than a block.

That's an issue particularly when it comes to tables. No matter which of the table formats you use, a reST table must always be constructed as a single block element. But when it's processed for translation, each table _cell_ will be a separately translatable string. There's no convenient way ReST could support embedded translator notes on individual table cells, because where would you even put them? Even what little context might exist — a column heading, for example — can't really be used in the general case. ReST tables can have 0, 1, or multiple header rows and/or header columns, and there's no way to automatically determine what's contextually relevant for a given cell.

#### Options (not many)
As far as I can tell, there's no way to embed any translator comments in `.rst` files ­— they don't get included in the `.pot` file.

Comments _can_ be **inserted** into the `.pot` file, and when `sphinx-intl` processes those masters into `.po` files, it will preserve them. But they have to be there _before_ the `.po` is generated — it doesn't consider a comment change or insertion to be an "update", so it won't regenerate a `.po` file from a `.pot` file that's only been edited to insert comments.

Worse, when the `.pot` files are regenerated (which has to be done any time the documentation is updated, to sync up the translations), the previous masters are overwritten. Any manually-inserted comments they contained will be **lost**. Regenerating `.po` files with `sphinx-intl` won't remove the comments if the `.pot` is otherwise _unchanged_ (comment removal isn't considered a change, either), but if there are **any** changes in a `.pot` file, the corresponding `.po` files will be re-generated and all comments will again be lost.

#### Another issue: Translation scope
There's also the issue that for regular prose content, the unit of translation is an entire **block**. While titles, table cells, etc. will be listed as individual units, the entire contents of a text block will instead be _combined_ into a single translatable string. Even if the file is written sentence-per-line (alternative term: [Semantic newlines](https://rhodesmill.org/brandon/2012/one-sentence-per-line/)), in the translation master the sentences will all be flowed together.

On one hand, that makes a good deal of sense from a translation perspective. The surrounding text often provides useful context, when trying to interpret a sentence's meaning. But it means that changing even a single word invalidates the translation of that entire paragraph, creating a lot of work for translators.

**Comment by ferdnyc** [comment]
The handling of gettext translations in ReST/Sphinx — the lack of comments, the table cell issue, the paragraph granularity — arguably makes it questionable whether it's even worth translating content that way. It may be simpler for writers to just translate entire _documents_ into their target language directly, rather than going through an intermediate file format like `.pot`/`.po`.

In part the answer to that question will depend on how often documentation is updated, and how sweeping those updates are.

Documentation that receives frequent updates which only change small portions of the content benefits greatly from using gettext translations, as only the changed portions will require updated translations. Most of the existing translations can be retained, as they'll still apply when translating the updated document.

When documentation is updated infrequently and en masse, the entire gettext translation will be invalidated. The translator will end up having to re-translate basically the entire text, but they'll be forced to do it in the somewhat-awkward chunks imposed by the `.po` file structure.  It's probably easier if they dispense with that, and just translate directly from the source.

Our documentation has to date been updated almost never, and updates have either made small changes across large portions of the source, or they've added entirely new chapters without altering the surrounding content at all. Neither is a perfect fit for either approach. In fact, they tend to favor opposite conclusions about which approach is best.

**Comment by ferdnyc** [comment]
(Trying to power through the rest of what's here, if I can stop myself from rambling on...)

> As for making images for every language, that would indeed be a lot of work. But you said before that all actions on images should be in the description anyway, so using the English image for a vague indication would probably work well enough.

Yeah, exactly. Images in English _should_ serve to illustrate interface descriptions, especially in the macro sense. The text won't match what someone using OpenShot in another language will see on their own screen, but that will only matter sometimes. 

(Some of our interface elements look kind of "samey", simply because they're built from the same Qt interface elements. Without their labels, it can be difficult to tell when you're looking at e.g. the Transitions list rather than the Project Files list or the Effects list. Others are text-heavy to a fault, so for example illustrating the Advanced settings tab of the Export dialog with a different interface language is unlikely to do any good. Which really just means there's no point in including a screenshot at all, when documenting that interface. A screenshot would add little value even in English.)

> (NB: still waiting the in docs topic for your actual permission to re-use your comments here in , as per your own statement that it should not be assumed. )

I will respond to that, but over there, to avoid discussions getting any more fractured. I really should've put the sentence-per-line stuff there as well, so that's on me.

> I've been leaving commenting in the docs with : : because it does not show up, would that work for this too?

Comments in ReST are started with two dots (`..`) — that's what you meant, because two colons (`::`) only has special meaning in three places: When they appear alone on a line by themselves, when they appear at the very end of a line, or when they're used as part of a directive in a line that starts with `..`. In other cases, they'll definitely show up in the output.

A line preceded by two periods, though, is _either_ a comment or the start of an explicit markup line (a ReST command, basically). It's an explicit markup line if the text that follows the `..` can be parsed as a ReST directive. If it can't be parsed, it's a comment. So, text preceded with `..` is definitely correct for comments... and also for special markup.

Either way it won't appear in the generated output, but that ambiguity of maybe being a directive, maybe being ignored, makes things kind of ugly. The specification [literally says](https://docutils.sourceforge.io/docs/ref/rst/restructuredtext.html#quick-syntax-overview), regarding comments preceded by `..`: "Anything may follow, except for the syntax of footnotes/citations, hyperlink targets, directives, or substitution definitions." (Because that syntax uses the same `..` initiator, and will be processed as a directive.)

Some Sphinx documentation suggests putting comments in an indented block _following_ the `..` line, to avoid the possibility of the comment text getting processed as a directive. Doing it that way, you have to indent every line of the comment, but only by one space. Probably not really worth it.

> I'll do a demo of it in my first complete update so we can see the difference.

The only thing I'd suggest, if you're doing updates, is to submit PRs as you go, section-by-section perhaps, rather than waiting for a "complete update". For a few reasons:
1. It lets you get a feel for the whole PR process, which like most things Git is fairly daunting for a first-timer. It's best to get that out of the way as early as possible, so you can develop a workflow for managing and submitting sets of changes that will work for both you and the PR process.
1. It gives the opportunity for feedback as early as possible.
1. It lets you react to that feedback as you continue on to the rest of the docs. If the feedback you get on a PR leads you to adjust your approach, you can incorporate that adjustment on updates to subsequent sections, rather than having to go back over everything you've already done and make changes to your changes.
1. It keeps the size of the PR more manageable, increasing the chances you'll get feedback from more people.
    (As you may have noticed, OpenShot accumulates [a fair number of PRs](https://github.com/OpenShot/openshot-qt/pulls), many of which are open for weeks or months with little to no response. Ideally every PR would be reviewed by multiple devs and/or other users, each providing constructive commentary on the changes — but that doesn't really happen. _Bigger_ PRs, in particular, struggle to get the attention of many of the devs, simply because they require more of a time commitment to look over the changes. A smaller, more self-contained PR has a better chance of being given a quick read-through.)
1. PRs can still be worked on while they're open. There's nothing stopping you from adding additional changes to a PR after it's already been submitted, just by committing those changes to the same branch of your fork. (A PR is opened from a particular branch, and it's recommended to create a separate branch for each logical group of changes. That keep those changes grouped together, and it also keeps them separated from other, unrelated changes. All of which goes back to the whole "getting a feel for the process" thing.)
    Point is, while a PR should be logically complete before it's submitted (so that, if it were immediately merged, it wouldn't result in the project repo containing files in an inconsistent or unfinished state), a PR doesn't become _inalterable_ upon submission. If you submit a PR that, say, adds image captions in one section, and the feedback you get is, "Nice, we should do this for all of the sections", those changes can be made and added to the existing PR.

And... \*phew\*, I think that's everything. At least, here in _this_ issue.

**Comment by MBB232** [comment]
Thanks for going above and beyond answering my questions. That was a bit more information then (I thought) I asked for. I hope I'm not distracting you from more urgent tasks. Very insightful, though a lot to take in. 

I agree with the importance of keeping relevant information together. 
It would be a shame if you had to rewrite all of this again for others. 
So I would propose you edit-cut-paste the parts to the other topic that you feel should be there. 
On the other hand, that thread is intended as call for volunteers, having it too long may be intimidating? 

Your comments can probably form a the basics of
documentation_About_Sphinx.rst 
and 
Documentation_About_Github.rst
Probably would not hurt to dumping  most of it into them for now, though they may need some reformatting and supplementation later. 

Yes, I meant two dots, not two colons [MBB goes to double check and fix ;-) ]

I am post-typewriter, but I also learned in school that rule about a space (single, not double) after a period in what may very well have been the last existing version of a Word Perfect office suite. (Called Disk-it, a file format compatible with literally nothing. I was happily adding them in before by <br> tags. I never realized that a linebrake could be seen as single space (Like I said, you wre very insightful in explaining things)

As for the translating tables thing; I meant that I understood them to be created by three lines of ==, where the number of == define the column widths - unless otherwise specified. 

====  ====  =======
Col1  Col2  Col3
====  ====  =======
1     One   Value 1
====  ====  =======

Translating to f.i. German would get 

====  ====  =======
Säule1  Säule2  Säule3
====  ====  =======
1     Eins   Wert 1
====  ====  =======

Which I guessed would cause issues when the translating words have a different length the the English
I thought that by explicitly defining the column widths as relative, this may be prevented.

**Comment by ferdnyc** [comment]
> Thanks for going above and beyond answering my questions. That was a bit more information then (I thought) I asked for.

Yeah, sorry. Some of that was me kind of journaling my thought process, as I looked into these things in real time, so not only is it excessive amounts of info but it isn't even consistent because my understanding of things was still evolving as I was writing about them.

And then some of it's just that I never seen to get tired of my own voice. But it's a perfectly reasonable defense mechanism to tune at _least_ 40% of it out.

> I hope I'm not distracting you from more urgent tasks. 
Believe me, there are no more urgent tasks than potentially bringing on a new contributor. I just hope I'm not completely overwhelming you. Just remember: <kbd>PgDn</kbd> and/or your mouse scroll are your friends. I am best appreciated at a brisk skim.

> Very insightful, though a lot to take in. 
Oh, then you'll really enjoy the _other_ issue! (I broke it up into conveniently-labeled sections! ...A LOT of sections. 😳)

> I agree with the importance of keeping relevant information together. 
> It would be a shame if you had to rewrite all of this again for others. 

Eh, it's GitHub. Cross referencing is only a link away.

> On the other hand, that thread is intended as call for volunteers, having it too long may be intimidating? 

That's certainly a consideration, but I view it as a selection mechanism. Because, if _that's_ enough to scare someone off, there's no chance they'd have survived me.


> As for the translating tables thing; 

Ah, gotcha! Yeah, translating content in a ReST document would be a disaster for exactly the reasons you brought up. Fortunately, they took an approach that avoids the issue entirely.

I'm not 100% sure where exactly translation is wired in with Sphinx, but I believe it works for any output format so I can guess.

Sphinx is a system for generating formatted docs from ReST source files. Depending how you have it configured, you feed in ReST files and HTML pages come out, or a LaTeX document comes out, or etc.

When translation is enabled, you feed in your same ReST source files in English (or whatever language they started in), _plus_ a `.mo` file (that's the compiled binary database form of a `.po` file), and depending how you have it configured, HTML pages in German come out, or a LaTeX document in Chinese comes out.

The ReST files only ever exist in their original form. The translations are applied to the parsed content during a step along the conversion pipeline, after it's already been processed from the ReST source and whitespace is no longer relevant. 

Clever! Just a shame about the whole comment/notes thing.

**Comment by MBB232** [comment]
> Eh, it's GitHub. Cross referencing is only a link away.

Yeah...that would explain why I now have 12 tabs about Openshot open in 3 browsers spread out over 2 computers... 
... The one that explained Github pull requests must be...somewhere... :-)
Does the Github Desktop version support this issue tracker too? because the comment box scales horribly in Chrome. )

> 
> Sure, I also added it to the docs-update Project. I pulled "bug", though, because to the extent these are openshot.org site bugs, those aren't tracked here. 😁 (They aren't tracked _anywhere_, formally, but point is there's no OpenShot application bug involved.)

Of course they are not. Why add just a new tag 'Website" when you can create a new separate bugtracker for it. :-D
Where Windows is slowly moving to single-sign-on, FOSS seems to really like to proliferate its logins, doesn't it? 
I think I have 3 accounts for LibreOffice, and other 3 for WinGPG and all I had were simple questions/bug reports. I assume actual programmers/contributors need even more accounts. 

I now have credentials on 3 platforms of Openshot, and would need a forth* if I wanted to do translations. 
At least you use quite common ones with Reddit, Disqus and Github. (But would backfire if people have different nicks on them. )
And there was expressed some desire on Reddit for a hosting platform of user-generated content (templates, title artwork etc) too. I'm sure it can get it's own platform with a separate login system. 
Does this not fracture the community? Or is it intended that way to prevent people from posting their questions in the bugtracker?

*and now you tell me there is a wiki too?

**Comment by ferdnyc** [comment]
> > Eh, it's GitHub. Cross referencing is only a link away.
> 
> Yeah...that would explain why I now have 12 tabs about Openshot open in 3 browsers spread out over 2 computers...

Hey, how you disorganize your life is entirely on you. :wink: Most times I typically have 30-50 tabs open in Chrome (though it can get up into the triple digits temporarily, when I'm plate-spinning), but I'm pathological. And I only have one pair of hands, so I tend to stick to one computer at a time. (Not counting my phone.)

> ... The one that explained Github pull requests must be...somewhere... :-)
> Does the Github Desktop version support this issue tracker too? because the comment box scales horribly in Chrome. )

It doesn't at all, no. It's purely a Git-based transaction-management tool, I guess is how I'd put it. You can send and retrieve commits, manage and switch between branches, etc., but that's it. Editing tasks, whether they involve the local files or the remote content / metacontent, require other tools.

I don't know specifically what scale issues you're experiencing, but Github in general has plenty of size/layout issues so it's not surprising. The fact that comment edit boxes start out so tiny drives me nuts, but at least they can usually be resized. (Though sometimes they'll fight growing longer because the page won't also extend to make room.) The fact that comment _width_ has a hard limit is mind-boggling, especially since it means forced horizontal scrolling whenever you embed a block of fixed-width text (including code snippets). The thinking behind that has always baffled me. But, since the width is fixed in their web interface, I assume any other tools would impose the same layout for the sake of consistency.

> > Sure, I also added it to the docs-update Project. I pulled "bug", though, because to the extent these are openshot.org site bugs, those aren't tracked here. grin (They aren't tracked _anywhere_, formally, but point is there's no OpenShot application bug involved.)
> 
> Of course they are not. Why add just a new tag 'Website" when you can create a new separate bugtracker for it. :-D

Well, because the website isn't managed here. Not only is the code not stored here, we don't even have access to the code.

My personal preference would be to keep the site code in a Github repo (a _separate_ one from openshot-qt, though), with its own issues list where website bugs could be tracked. But since I don't run the website, it's not up to me. And as long as the code isn't available, there's no point in publicly tracking site bugs.

> I now have credentials on 3 platforms of Openshot, and would need a forth* if I wanted to do translations.
> At least you use quite common ones with Reddit, Disqus and Github. (But would backfire if people have different nicks on them. )
> And there was expressed some desire on Reddit for a hosting platform of user-generated content (templates, title artwork etc) too. I'm sure it can get it's own platform with a separate login system.
> Does this not fracture the community?

`¯\_(ツ)_/¯` We consolidate what we can. If translations could be managed on Github, I'm sure we'd at least consider moving them here, but that's not an available feature. We could move translations from launchpad to a dedicated site for crowd-sourced translations — Crowdin, for example — but that'd still be **another** account people would have to open.

> Or is it intended that way to prevent people from posting their questions in the bugtracker?

Things like the Subreddit are there because there's a demand, primarily. Discussions there reach a much larger community of active users than posts in our issue tracker do. So it's a better place to ask questions and get a timelier answer. I don't personally ever look at it, even though I _have_ a Reddit account, because I've more than got my hands full already just with the traffic here.

> *and now you tell me there is a wiki too?

Well, yeah but that's part of the Github repo:
![image](https://user-images.githubusercontent.com/538020/80673452-49af6d00-8a7d-11ea-92c3-47ba4935457f.png)

No additional logins required.

**Comment by ferdnyc** [comment]
On the strength of this report, plus #3475 by @yfdyh000, @jonoomph took another run at how the website manages content language selection. Hopefully it's more visible and accessible now.

**Comment by MBB232** [comment]
@ferdnyc @jonoomph 
What, by adding it to the very bottom of the page? I did not know I could scroll that low 😎
Can't it just be moved up in the menu?

![Language_Screenshot_marked](https://user-images.githubusercontent.com/62191951/82147262-dc5a5500-984e-11ea-98e9-9bfd330a2c45.png)

**Comment by ferdnyc** [comment]
### \*COUGH\*

![82147262-dc5a5500-984e-11ea-98e9-9bfd330a2c45](https://user-images.githubusercontent.com/538020/82147767-fd568200-981d-11ea-8bc8-ca33851ffce3.png)

(See @jonoomph? I said it blended in too well with the social icons.)

**Comment by ferdnyc** [comment]
Also, I gather there's an **additional** language-switching bar/banner that shows up at the top of the page, but only when you're browsing the site in a language that doesn't match your browser's... or something like that.

Ah, yes, it's actually at the top of the content, just below the menu (when scrolled to the top).
![image](https://user-images.githubusercontent.com/538020/82147871-c03ebf80-981e-11ea-82e6-bfa9dda58842.png)
